### PR TITLE
StableDiffusionLatentUpscalePipeline - positive/negative prompt embeds support

### LINKED ISSUE #8895: StableDiffusionLatentUpscalePipeline - positive/negative prompt embeds support
I'm trying to deploy the smallest possible SD inpainting model. My production deployment only needs unet+vae+ipadapter weights with prompt and ip adapter image embeds pre-generated. Works well!

Now I wanted to try latent upscaler from diffusers and realized it currently doesn't support pre-generated embeds. Would probably be nice to keep its API aligned with the rest and add them.

**Describe the solution you'd like.**

Harmonizing inputs on the StableDiffusionLatentUpscalePipeline with other more frequently used pipelines would be nice.

### PR DESCRIPTION
**Pull request by rootonchair**
# What does this PR do?

Fixes #8895

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu

### CODE REVIEW

**Comment by rootonchair** [comment]
Test script
```python
from diffusers import StableDiffusionLatentUpscalePipeline, StableDiffusionPipeline
import torch

pipeline = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16)
pipeline.to("cuda")

upscaler = StableDiffusionLatentUpscalePipeline.from_pretrained("stabilityai/sd-x2-latent-upscaler", torch_dtype=torch.float16)
upscaler.to("cuda")

prompt = "a photo of an astronaut high resolution, unreal engine, ultra realistic"
generator = torch.manual_seed(33)

# we stay in latent space! Let's make sure that Stable Diffusion returns the image
# in latent space
low_res_latents = pipeline(prompt, generator=generator, output_type="latent").images

upscaled_image = upscaler(
    prompt=prompt,
    image=low_res_latents,
    num_inference_steps=20,
    guidance_scale=0,
    generator=generator,
).images[0]

# Let's save the upscaled image under "upscaled_astronaut.png"
upscaled_image.save("astronaut_1024.png")

prompt_embeds, negative_prompt_embeds, pooled_prompt_embeds, negative_pooled_prompt_embeds = upscaler.encode_prompt(
    prompt=prompt,
    device=upscaler._execution_device,
    do_classifier_free_guidance=False,
)

upscaled_image = upscaler(
    image=low_res_latents,
    num_inference_steps=20,
    guidance_scale=0,
    generator=generator,
    prompt_embeds=prompt_embeds,
    pooled_prompt_embeds=pooled_prompt_embeds,
).images[0]

upscaled_image.save("embeds_astronaut_1024.png")

# as a comparison: Let's also save the low-res image
with torch.no_grad():
    image = pipeline.decode_latents(low_res_latents)
image = pipeline.numpy_to_pil(image)[0]

image.save("astronaut_512.png")
```

**Review by yiyixuxu — APPROVED** [review]
oh thanks! I think it looks really nice!

**Review by sayakpaul — COMMENTED** [review]
LGTM! Could you also share some results stemming from this change? 

Let's also add a fast test for this?

**Comment by rootonchair** [comment]
> Could you also share some results stemming from this change?

Sure
Here is the original result
org_astronaut_1024



And result using embeds
astronaut_1024

which does not have much different

> Let's also add a fast test for this?

Should I open another PR for this @sayakpaul ?

**Comment by rootonchair** [comment]
Hi @sayakpaul, could you help me review this PR?

File: tests/pipelines/stable_diffusion/test_stable_diffusion_latent_upscaler.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
+# coding=utf-8
```
Comment (yiyixuxu) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_latent_upscaler.py:1]: oh actually I think the tests for latent upscaler is here https://github.com/huggingface/diffusers/blob/main/tests/pipelines/stable_diffusion_2/test_stable_diffusion_latent_upscale.py

maybe we add new tests there? very sorry for all the additional work to create a new test from scratch

**Comment by rootonchair** [comment]
@yiyixuxu I have just update the existing test. Could you help me review? Btw, adding new test is kinda interesting too. So no worries

File: tests/pipelines/stable_diffusion_2/test_stable_diffusion_latent_upscale.py [hunk 2]
```diff
@@ -173,11 +173,51 @@ def test_inference(self):
 
         self.assertEqual(image.shape, (1, 256, 256, 3))
         expected_slice = np.array(
-            [0.47222412, 0.41921633, 0.44717434, 0.46874192, 0.42588258, 0.46150726, 0.4677534, 0.45583832, 0.48579055]
+            [0.3970313, 0.3768756, 0.41147298, 0.4716793, 0.5115408, 0.44601366, 0.43763855, 0.46781355, 0.46358708]
```
Comment (yiyixuxu) [inline tests/pipelines/stable_diffusion_2/test_stable_diffusion_latent_upscale.py:176]: why do we need to update the expected_slice here? the results of existing tests should not change, no?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py [hunk 3]
```diff
@@ -373,16 +517,32 @@ def __call__(
...
+            pooled_prompt_embeds = torch.cat([pooled_prompt_embeds, negative_pooled_prompt_embeds])
+
         # 4. Preprocess image
         image = self.image_processor.preprocess(image)
-        image = image.to(dtype=text_embeddings.dtype, device=device)
+        image = image.to(dtype=prompt_embeds.dtype, device=device)
         if image.shape[1] == 3:
             # encode image if not in latent-space yet
-            image = self.vae.encode(image).latent_dist.sample() * self.vae.config.scaling_factor
+            image = retrieve_latents(self.vae.encode(image), generator=generator) * self.vae.config.scaling_factor
```
Comment (rootonchair) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py:545]: @yiyixuxu I think it's due to this line. The old code does not take in generator

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py [hunk 4]
```diff
@@ -373,16 +517,32 @@ def __call__(
...
+            negative_prompt,
+            prompt_embeds,
+            negative_prompt_embeds,
+            pooled_prompt_embeds,
+            negative_pooled_prompt_embeds,
         )
 
+        if do_classifier_free_guidance:
+            prompt_embeds = torch.cat([prompt_embeds, negative_prompt_embeds])
+            pooled_prompt_embeds = torch.cat([pooled_prompt_embeds, negative_pooled_prompt_embeds])
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py:538]:
```suggestion
            prompt_embeds = torch.cat([negative_prompt_embeds, prompt_embeds])
            pooled_prompt_embeds = torch.cat([negative_pooled_prompt_embeds, pooled_prompt_embeds])
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py [hunk 5]
```diff
@@ -105,7 +119,54 @@ def __init__(
...
+            negative_prompt=negative_prompt,
+            prompt_embeds=prompt_embeds,
+            negative_prompt_embeds=negative_prompt_embeds,
+            pooled_prompt_embeds=pooled_prompt_embeds,
+            negative_pooled_prompt_embeds=negative_pooled_prompt_embeds,
+            **kwargs,
+        )
+
+        prompt_embeds = torch.cat([prompt_embeds, negative_prompt_embeds])
+        pooled_prompt_embeds = torch.cat([pooled_prompt_embeds, negative_pooled_prompt_embeds])
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py:155]:
```suggestion
        prompt_embeds = torch.cat([negative_prompt_embeds, prompt_embeds])
        pooled_prompt_embeds = torch.cat([negative_pooled_prompt_embeds, pooled_prompt_embeds])
```

File: tests/pipelines/stable_diffusion_2/test_stable_diffusion_latent_upscale.py (hunk 2 above)
Comment (yiyixuxu) [inline tests/pipelines/stable_diffusion_2/test_stable_diffusion_latent_upscale.py:176]:
```suggestion
            [0.47222412, 0.41921633, 0.44717434, 0.46874192, 0.42588258, 0.46150726, 0.4677534, 0.45583832, 0.48579055]
```

**Review by yiyixuxu — COMMENTED** [review]
I think we got the order of the negative_prompt_embeds and prompt_embeds reversed, that's why the previous test wasn't passing. Let's make the change, and change the test back and make sure it passes :)

**Comment by rootonchair** [comment]
@yiyixuxu thanks for the catch! Sorry I did not check it thoroughly

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Pull request merged**